Add tests and restrict !user action to authorized users

dawaa opened the pull request:
This PR became somewhat bigger than I first anticipated, sorry about that.

If necessary I could split the initial set up of the tests and mocks into its own PR and create a second one that only holds the relevant changes to restrict `!user` action that'd be based off of the first one.

Because of my lack of knowledge regarding both IRC and Hubot I had to really dig around to get an understanding of what was going on, hence why I ended up creating the tests and all other things around that.

#### Goal

Was to restrict `!user` action from being run by unauthorized users.

proton-ab commented:
Doesn't seem to be working properly, !user command is still accepted by anyone on IRC.

proton-ab commented on scripts/animebytes.coffee:
```diff
@@ -61,7 +64,11 @@ module.exports = (robot) ->
     room = msg.message.room
-    name = msg.match[1] || parseUserInfo(msg.message.user.original.host).user
+    try
+      name = msg.match[1] || parseUserInfo(msg.message.user.original.host).user
```
Node, as any other language, optimizes execution flow, hence if msg.match[1] satisfies simple '||' expression, it will forgo calling parseUserInfo (which is pointless at this point). Due to that, if unauthorized user calls '!user test', code will never have a chance of throwing exception. Should you change order in this expression, then parseUserInfo will always be true for authorized users and parameters of !user will never work.

proton-ab commented:
Additionally, it would seem that I am unable to execute tests for animebytes.test.coffee, resulting in

```
sakuya@chat:~/kana$ ./node_modules/mocha/bin/mocha -- ./test/animebytes.test.coffee 


  0 passing (2ms)
```

parseUserInfo test seems to be the only one working

```
./node_modules/mocha/bin/mocha -- ./test/*.coffee


  parseUserInfo utility
    when valid input
      ✓ should parse user info
      ✓ should throw on wrong host
    when invalid input
      ✓ should throw on invalid input
      ✓ should throw on missing user
      ✓ should throw on missing rank
      ✓ should throw on missing host


  6 passing (27ms)
```

You should definitely include variation of !user with and without parameters to parseUserInfo test.

dawaa commented:
@proton-ab I've updated my PR with 1 extra commit (I noticed the `room` variable was hanging) and a `fixup`-commit that should not allow unauthorized users to call `!user [name]`.

___________________________________________

### Not able to run tests

Regarding not being able to run the tests, my fault I should've been more clear about it initially. In https://github.com/anniemaybytes/kana/pull/2/commits/44b06a72764a364d3981aeee86b1ac9a1ecc3678#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R7 I added two commands in the `package.json`. Please try again using `npm run test:all`, I'm also making use of the [mocha.opts](https://github.com/anniemaybytes/kana/blob/44b06a72764a364d3981aeee86b1ac9a1ecc3678/test/mocha.opts) file to pass a couple of arguments to `mocha`.

___________________________________________

##### `"You should definitely include variation of !user with and without parameters to parseUserInfo test."`
I'm not sure I follow you on this one, what `parseUserInfo` takes is just a string, similar to something like `123@dawaa.User.SomeHost`.

However if I understand you correctly, I already have those tests. [Calling without a given user](https://github.com/dawaa/kana/commit/da3ae0b052fd3d744aa11f6574f0b2113272e7db#diff-2f362d19c3b05086346753dbb1f1ebeaR82) and [when passing a user](https://github.com/dawaa/kana/commit/da3ae0b052fd3d744aa11f6574f0b2113272e7db#diff-2f362d19c3b05086346753dbb1f1ebeaR94)

dawaa commented:
@proton-ab Thanks, changed from using default port and anything within [the usual range between ` 6660–6669` and  `7000`](https://en.wikipedia.org/wiki/Internet_Relay_Chat#cite_ref-24).

proton-ab commented:
All looks good now.

proton-ab commented:
https://github.com/anniemaybytes/kana/commit/7469f72c5b0cf0a7b4181eba0d50c299cf0b1530
